### PR TITLE
[FEATURE] [needs-docs] Metadata editor

### PR DESCRIPTION
**Pull request by Gustry**
## Description

Part of the QEP 91 https://github.com/qgis/QGIS-Enhancement-Proposals/issues/91

This PR is for the metadata editor.
It adds a new entry in the layer properties, because we want to keep the existing metadata panel until this QEP is finished. According to a discussion with @nyalldawson, we will do the cleaning between old and new API after Work Package 3 (Metadata Storage)

This is for both rasters and vectors.

screen shot 2017-08-17 at 11 56 16
screen shot 2017-08-17 at 11 56 39
screen shot 2017-08-17 at 11 56 51
screen shot 2017-08-17 at 11 57 04

I will add Contacts and Extents (temporal and spatial) in a separate PR. In the mean time, I would like comments and get this one merge if possible.

Funded by World Bank

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

CC: @timlinux

### CODE REVIEW

**Comment by Gustry** [comment]
Hi devs,
Is-it possible to have a review here?
Thanks !

**Comment by timlinux** [comment]
@nyalldawson @m-kuhn @wonder-sk or anyone else interested - could you take a look at this PR?

File: src/core/qgsapplication.cpp [hunk 1]
```diff
@@ -623,6 +623,11 @@ QString QgsApplication::i18nPath()
     return ABISYM( mPkgDataPath ) + QStringLiteral( "/i18n/" );
 }
 
+QString QgsApplication::metadataPath()
```
Comment (m-kuhn) [inline src/core/qgsapplication.cpp]: const

File: src/gui/qgsmetadatawidget.cpp [hunk 2]
```diff
@@ -0,0 +1,708 @@
...
+  fillComboBox();
+  setPropertiesFromLayer();
+  updateContactDetails();
+}
+
+QgsMetadataWidget::~QgsMetadataWidget()
+{
+}
+
+void QgsMetadataWidget::setAutoSource() const
```
Comment (m-kuhn) [inline src/gui/qgsmetadatawidget.cpp]: Setters are normally not const, is this line correct?

File: src/core/qgsapplication.cpp (hunk 1 above)
Comment (Gustry) [reply src/core/qgsapplication.cpp]: This one is `static` in the header already

File: src/gui/qgsmetadatawidget.cpp (hunk 2 above)
Comment (Gustry) [reply src/gui/qgsmetadatawidget.cpp]: Sorry, it's misleading, it's not a setter. It's a function to set text automatically in a QLineEdit according to the layer
screen shot 2017-09-14 at 11 28 49

**Comment by timlinux** [comment]
@nyalldawson could you take a peek at this - we would like to get it merged tomorrow if possible....

Thanks!

**Comment by Gustry** [comment]
Tim, I know I have I will have some issues.
And I'm waiting for https://github.com/qgis/QGIS/pull/5193 to be merged first. Then I will need to rebase

File: src/gui/qgsmetadatawidget.cpp [hunk 3]
```diff
@@ -0,0 +1,806 @@
...
+  return results;
+}
+
+QMap<QString, QString> QgsMetadataWidget::parseLanguages()
+{
+  QString path = QDir( QgsApplication::metadataPath() ).absoluteFilePath( QString( "country_code_ISO_3166.csv" ) );
+  QFile file( path );
+  if ( !file.open( QIODevice::ReadOnly ) )
+  {
+    QgsDebugMsg( QString( "Error while opening the CSV file: %1, %2 " ).arg( path, file.errorString() ) );
```
Comment (nyalldawson) [inline src/gui/qgsmetadatawidget.cpp]: shouldn't this return here?

File: src/gui/qgsmetadatawidget.cpp [hunk 4]
```diff
@@ -0,0 +1,806 @@
...
+  }
```
Comment (nyalldawson) [inline src/gui/qgsmetadatawidget.cpp]: Shouldn't this return here? (also in the other parsing functions)

File: src/gui/qgsmetadatawidget.cpp [hunk 5]
```diff
@@ -0,0 +1,806 @@
...
+{
+  QString path = QDir( QgsApplication::metadataPath() ).absoluteFilePath( QString( "md_scope_codes.csv" ) );
+  QFile file( path );
+  if ( !file.open( QIODevice::ReadOnly ) )
+  {
+    QgsDebugMsg( QString( "Error while opening the CSV file: %1, %2 " ).arg( path, file.errorString() ) );
+  }
+
+  QMap<QString, QString> types;
+  types.insert( "", "" ); // We add an empty line, because it's not compulsory.
```
Comment (nyalldawson) [inline src/gui/qgsmetadatawidget.cpp]: "" - > QString()

File: src/gui/qgsmetadatawidget.cpp [hunk 6]
```diff
@@ -0,0 +1,806 @@
...
+  QString currentTabText = tabWidget->widget( index )->objectName();
+  if ( currentTabText == "tabCategoriesDialog" )
+  {
+    // Categories tab
+    // We need to take keywords and insert them into the list
+    QList<QTableWidgetItem *> categories = tabKeywords->findItems( "gmd:topicCategory", Qt::MatchExactly );
+    if ( !categories.isEmpty() )
+    {
+      int row = categories.at( 0 )->row();
+      mCategoriesModel->setStringList( tabKeywords->item( row, 1 )->text().split( "," ) );
```
Comment (nyalldawson) [inline src/gui/qgsmetadatawidget.cpp]: .split( ',' ) (char based method is much faster)

File: src/gui/qgsmetadatawidget.cpp [hunk 7]
```diff
@@ -0,0 +1,806 @@
...
+    }
+    else
+    {
+      mCategoriesModel->setStringList( QStringList() );
+    }
+  }
+  else if ( currentTabText == "tabKeywordsDialog" )
```
Comment (nyalldawson) [inline src/gui/qgsmetadatawidget.cpp]: Lots of string literals in this file should be changed to QStringLiteral( ... )

File: src/gui/qgsmetadatawidget.cpp [hunk 8]
```diff
@@ -0,0 +1,806 @@
...
+    checkMetadata();
+  }
+}
+
+void QgsMetadataWidget::addNewCategory()
+{
+  bool ok;
+  QString text = QInputDialog::getText( this, tr( "New Category" ),
+                                        tr( "New Category:" ), QLineEdit::Normal,
+                                        QString( "" ), &ok );
```
Comment (nyalldawson) [inline src/gui/qgsmetadatawidget.cpp]: QString( "" ) -> QString()

File: src/gui/qgsmetadatawidget.h [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+    /**
+     * Add a new vocabulary.
+     */
+    void addVocabulary() const;
+
+    /**
+     * Remove a selected vocabulary.
+     */
+    void removeVocabulary() const;
```
Comment (nyalldawson) [inline src/gui/qgsmetadatawidget.h]: I think for clarity these should be renamed "removeSelectedVocabularies"/ etc

File: src/gui/qgsmetadatawidget.h [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+
+  public:
+
+    /**
+     * Constructor for the wizard.
+     */
+    QgsMetadataWidget( QWidget *parent, QgsMapLayer *layer = nullptr );
+    ~QgsMetadataWidget();
+
+    /**
```
Comment (nyalldawson) [inline src/gui/qgsmetadatawidget.h]: I think all these methods should be made private - we don't want to expose things like this to the API if there's no requirement for them to be used externally, and in this case I worry that locking these into the API might prove restrictive in future.

File: src/gui/qgsmetadatawidget.h [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+    QStringListModel *mCategoriesModel = nullptr;
+    QStringListModel *mDefaultCategoriesModel = nullptr;
+    QStringListModel *mRightsModel = nullptr;
+    QStringListModel *mHistoryModel = nullptr;
+    void syncFromCategoriesTabToKeywordsTab() const;
+};
+
+#ifndef SIP_RUN
+
+/**
```
Comment (nyalldawson) [inline src/gui/qgsmetadatawidget.h]: Also put a `\\\@cond PRIVATE` here, and `\\\@endcond` at the end of the file

This hides the classes from the doxygen files, making them truely private and not part of stable api

**Comment by nyalldawson** [comment]
Looks good! No major issues here, just a bunch of small cleanups

**Comment by timlinux** [comment]
Thanks so much for your review @nyalldawson

File: src/app/qgsrasterlayerproperties.h [hunk 12]
```diff
@@ -22,6 +22,7 @@
 #include "qgsoptionsdialogbase.h"
 #include "ui_qgsrasterlayerpropertiesbase.h"
 #include "qgsguiutils.h"
+#include "qgsmetadatawidget.h"
```
[the remark this answers is not in the document]
Comment (Gustry) [reply src/app/qgsrasterlayerproperties.h]: What do you mean by forward declared? You mean to add `class QgsMetadataWidget
` just below?

[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/app/qgsrasterlayerproperties.h]: You add the `class QgsMetadataWidget` in the header, and move the `#include "qgsmetadatawidget.h"` to the .cpp file. Wherever possible this should be done as it speeds up compilation.

File: src/app/qgsvectorlayerproperties.cpp [hunk 13]
```diff
@@ -247,6 +247,14 @@ QgsVectorLayerProperties::QgsVectorLayerProperties(
   diagLayout->addWidget( diagramPropertiesDialog );
   mDiagramFrame->setLayout( diagLayout );
 
+  // Metadata tab
+  QVBoxLayout *metadataLayout = new QVBoxLayout( metadataFrame );
+  metadataLayout->setMargin( 0 );
+  mMetadataWidget = new QgsMetadataWidget( this, mLayer );
+  mMetadataWidget->layout()->setContentsMargins( -1, 0, -1, 0 );
```
[the remark this answers is not in the document]
Comment (Gustry) [reply src/app/qgsvectorlayerproperties.cpp]: it's a copy/paste from all other widgets added in the window: https://github.com/qgis/QGIS/blob/master/src/app/qgsvectorlayerproperties.cpp#L246

I can have a look

[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/app/qgsvectorlayerproperties.cpp]: Ok - don't stress then. It's probably been like that forever and there's been no issue.

File: src/gui/qgsmetadatawidget.h (hunk 10 above)
Comment (Gustry) [reply src/gui/qgsmetadatawidget.h]: I agree I'm not sure what should be public. I'm not sure functions related to UI are very useful in the public section.
I moved most of them. Should I move more of them into the private section?

File: src/gui/qgsmetadatawidget.cpp (hunk 3 above)
Comment (Gustry) [reply src/gui/qgsmetadatawidget.cpp]: I'm returning an empty QStringList or QMap. Or Are you talking about the error message?

File: src/gui/qgsmetadatawidget.cpp [hunk 14]
```diff
@@ -0,0 +1,806 @@
...
+  {
+    struct QgsLayerMetadata::Constraint constraint = QgsLayerMetadata::Constraint();
+    constraint.type = mConstraintsModel->item( row, 0 )->text();
+    constraint.constraint = mConstraintsModel->item( row, 1 )->text();
+    constraints.append( constraint );
+  }
+  layerMetadata.setConstraints( constraints );
+
+  // CRS
+  if ( selectionCrs->crs().isValid() )
```
[the remark this answers is not in the document]
Comment (Gustry) [reply src/gui/qgsmetadatawidget.cpp]: I was not sure about the behaviour here if we generate a wrong XML for the CRS item or if something could be wrong later.

File: src/gui/qgsmetadatawidget.h (hunk 10 above)
Comment (Gustry) [reply src/gui/qgsmetadatawidget.h]: Two parsers are used in the `QStyledItemDelegate`. Can I let them public? They doesn't break anything I guess, they are just CSV parser.

**Comment by Gustry** [comment]
Ok, thanks @nyalldawson and @m-kuhn for the review. Promise, I will read docs about transfer ownership etc ;-)

Is-there any comments to fix?

I know the UI is not the best, but we can still improve it incrementally if needed.

**Comment by Gustry** [comment]
Travis is failing on the segmentation fault from the server side

**Comment by rldhont** [comment]
@m-kuhn is working on fixing the segfault: https://github.com/qgis/QGIS/pull/5230

**Comment by timlinux** [comment]
Anything left to do here @Gustry ? Please as @nyalldawson to apply when you are ready...

File: src/gui/qgsmetadatawidget.h (hunk 10 above)
Comment (nyalldawson) [reply src/gui/qgsmetadatawidget.h]: Sure - leave anything public which needs to be, but hide everything which is just implementation details...

**Comment by Gustry** [comment]
I just fixed a conflict on the UI file, I think I could fix most of the comments.

**Comment by nyalldawson** [comment]
Looks to me like it's all been addressed, thanks!

**Comment by timlinux** [comment]
Thanks very much @nyalldawson  and @Gustry !

**Comment by kalxas** [comment]
Thanks!

**Pull request merged**